main/lsp/requests/rename.cc

```diff
@@ -224,6 +224,55 @@ class ConstRenamer : public Renamer {
     }
 };
 
+// Add subclass-related methods (methods overriding and overridden by `symbol`) to the `methods` vector.
+void addSubclassRelatedMethods(LSPTypecheckerDelegate &typechecker, core::SymbolRef symbol,
+                               vector<core::SymbolRef> &methods) {
+    const core::GlobalState &gs = typechecker.state();
+    auto symbolData = symbol.data(gs);
+
+    // We have to check for methods as part of a class hierarchy: Follow superClass() links till we find the root;
+    // then find the full tree; then look for methods with the same name as ours; then find all references to all
+    // those methods and rename them.
+    auto symbolClass = symbolData->enclosingClass(gs);
+
+    // We have to be careful to follow superclass links only as long as we find a method that `symbol` overrides.
+    // Otherwise we will find unrelated methods and rename them even though they don't need to be (see the
+    // method_class_hierarchy test case for an example).
+    auto root = findRootClassWithMethod(gs, symbolClass, symbolData->name);
+
+    auto subclasses = getSubclasses(typechecker, root);
+
+    // find the target method definition in each subclass
+    for (auto c : subclasses) {
+        auto classSymbol = c.data(gs);
+        auto member = classSymbol->findMember(gs, symbolData->name);
+        if (!member.exists()) {
+            continue;
+        }
+        methods.push_back(member);
+    }
+}
+
+// Add methods that are related because of dispatching via secondary components in sends (union types).
+void addDispatchRelatedMethods(LSPTypecheckerDelegate &typechecker, const core::DispatchResult *dispatchResult,
+                               vector<core::SymbolRef> &methods) {
+    for (const core::DispatchResult *dr = dispatchResult; dr; dr = dr->secondary.get()) {
+        auto method = dr->main.method;
+        ENFORCE(method.exists());
+        auto included = false;
+        for (auto m : methods) {
+            if (m == method) {
+                included = true;
+                break;
+            }
+        }
+        if (!included) {
+            // this also adds `method` itself
+            addSubclassRelatedMethods(typechecker, method, methods);
```

Comment:
`addSubclassRelatedMethods` doesn't deduplicate. I'm worried about footguns here with respect to forgetting to dedupe in various places as we build out this functionality. I wonder if we should have a simple wrapper around the vector for this use case.